src/CourseService.php

```diff
@@ -386,6 +386,89 @@ public function UpdateAssetsFromUploadedFile($courseId, $uploadLocation)
+    /// <returns>List of Import Results</returns>
+    public function ImportCourseAsync($courseId, $absoluteFilePathToZip)
+    {
+        $uploadService = new UploadService($this->_configuration);
+        $location = $uploadService->UploadFile($absoluteFilePathToZip);
+
+        $importException = null;
+        $response = null;
+        try {
+            $token = $this->ImportCourseAsyncToken($courseId, $location);
+            $done = false;
+            while (!$done){
+                $xmlStatus = $this->getAsyncImportStatus($token);
+                $statusResult = (string) $xmlStatus->status;
+                switch ($statusResult) {
+                    case 'running':
+                        sleep(5);
+                        break;
+                    case 'error':
+                        $done = true;
+                        $response = [];
+                        break;
+                    case 'finished' :
+                        $done = true;
+                        $importResult = new ImportResult(null);
+                        $response = $importResult->ConverToImportResultsFromXML($xmlStatus);
+                        break;
+                    default:
+                        $done = true;
+                        $response = [];
+                        break;
+                }
+            }
+        } catch (Exception $ex) {
+            $importException = $ex;
+        }
+        $uploadService->DeleteFile($location);
+        if($importException != null){
+            throw $importException;
+        }
+        return $response;
+    }
+
+    /// <summary>
+    /// Get import course token
+    /// </summary>
+    /// <param name="courseId">Unique Identifier for the course</param>
+    /// <param name="path">The relative path to the uploaded zip file</param>
+    /// <returns>Token Id</returns>
+    public function ImportCourseAsyncToken($courseId, $path)
+    {
+        $request = new ServiceRequest($this->_configuration);
+        $params = array('courseid'=>$courseId,
+            'path'=>$path);
+        $request->setMethodParams($params);
+        $response = $request->CallService("rustici.course.importCourseAsync");
+        write_log('rustici.course.importCourseAsync : '.$response);
+        $xmlToken = simplexml_load_string($response);
+        $tokenResult = (string) $xmlToken->token->id;
+        return $tokenResult;
+    }
+
+    /// <summary>
+    /// Get import course status and result using async client
+    /// </summary>
+    /// <param name="token">Unique token for the course upload</param>
+    /// <returns>Import Status and Result as SimpleXMLElement</returns>
+    private function getAsyncImportStatus($token)
+    {
+        $request = new ServiceRequest($this->_configuration);
+        $params = array('token'=>$token);
+        $request->setMethodParams($params);
+        $response = $request->CallService("rustici.course.getAsyncImportResult");
```

Comment:
is this a method we are calling from the library or somehow attaching to the library?

Reply:
The method is attached in the library. Here: https://github.com/GrovoLearning/SCORMCloud_PHPLibrary/blob/master/src/ServiceRequest.php. But the whole library we forked from someone else'

Reply:
Existing deployed code isn't using async approach, it is using sync approach, from line 61-80 and line 98-118 here https://github.com/GrovoLearning/SCORMCloud_PHPLibrary/blob/master/src/CourseService.php